instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go

```diff
@@ -152,6 +152,16 @@ func (c *config) handleRPC(ctx context.Context, rs stats.RPCStats, isServer bool
 	case *stats.InPayload:
 		if gctx != nil {
 			messageId = atomic.AddInt64(&gctx.inMessages, 1)
+
+			// If a MetricAttributesFn is defined call this function and update the gRPCContext with the metric attributes
+			// returned by ths function.
+			if f := c.MetricAttributesFn; f != nil {
+				attrs := f(ctx, rs.Payload)
+
+				gctx.metricAttrs = append(gctx.metricAttrs, attrs...)
```

Comment:
It doesn't seem like we need to modify `gctx` here?
This also causes a race problem.

Reply:
So the #6092  also modified this, otherwise the metric attributes won't be propagated on the `*stats.End` call where the majority of the metrics are recorded, since I don't think `*stats.End` has access to the Payload although I will check that.

Reply:
[`InPayload`](https://pkg.go.dev/google.golang.org/grpc/stats#InPayload) is the only `stats.RPCStats` implementation that has access to the request payload. So the this is only time we can call the `MetricAttributesFn` to be able to append too `gRPCContext.metricAttrs`, I can put a lock around  `metricAttrs` protect against races?

Reply:
I'm very hesitant about updating a pointer to struct that is stored as a context value. That really looks like a smell.
Unfortunately, we can't do things cleanly (provide a new context with the updated value), as we can't provide a new context as return value.

Reply:
Ok, this doesn't seem like it's going to be possible then due to the limitations in how the `stats.Handler` works.

Reply:
@dmathieu any thoughts on if we should continue with this or not?

Reply:
I'd like to have other opinions.

Reply:
Ok, I'll leave it with you @dmathieu :+1:

Reply:
Any updates on this?